pallets/circuit/src/lib.rs

```diff
@@ -648,62 +563,55 @@ pub mod pallet {
 
         #[pallet::weight(<T as pallet::Config>::WeightInfo::cancel_xtx())]
         pub fn cancel_xtx(origin: OriginFor<T>, xtx_id: T::Hash) -> DispatchResultWithPostInfo {
-            let requester = Self::authorize(origin, CircuitRole::Requester)?;
-            // Setup: new xtx context
-            let mut local_ctx: LocalXtxCtx<T> =
-                Self::setup(CircuitStatus::PendingBidding, &requester, Some(xtx_id))?;
-
-            if requester != local_ctx.xtx.requester
-                || local_ctx.xtx.status > CircuitStatus::PendingBidding
-            {
-                return Err(Error::<T>::UnauthorizedCancellation.into())
-            }
+            let attempting_requester = Self::authorize(origin, CircuitRole::Requester)?;
 
-            // Drop cancellation in case some bids have already been posted
-            if Self::get_current_step_fsx(&local_ctx)
-                .iter()
-                .any(|fsx| fsx.best_bid.is_some())
-            {
-                return Err(Error::<T>::UnauthorizedCancellation.into())
-            }
+            Machine::<T>::compile(
+                &mut Machine::<T>::load_xtx(xtx_id)?,
+                |current_fsx, _local_state, _steps_cnt, status, requester| {
+                    if attempting_requester != requester || status > CircuitStatus::PendingBidding {
+                        return Err(Error::<T>::UnauthorizedCancellation)
+                    }
+                    // Drop cancellation in case some bids have already been posted
+                    if current_fsx.iter().any(|fsx| fsx.best_bid.is_some()) {
+                        return Err(Error::<T>::UnauthorizedCancellation)
+                    }
+                    Ok(PrecompileResult::TryKill(Cause::IntentionalKill))
+                },
+                no_post_updates,
+            )?;
 
-            Self::kill(&mut local_ctx, CircuitStatus::DroppedAtBidding);
+            Ok(().into())
+        }
 
+        #[pallet::weight(<T as pallet::Config>::WeightInfo::cancel_xtx())]
+        pub fn revert(origin: OriginFor<T>, xtx_id: T::Hash) -> DispatchResultWithPostInfo {
```

Comment:
Is the correct transition enforced here, ensuring that root can't force illegal transitions?

Reply:
as long as revert comes before FinishedAllSteps it will go thru. Can cover with test